src/Stack/Fetch.hs

```diff
@@ -554,6 +533,69 @@ fetchPackages' mdistDir toFetchAll = do
                 S.writeFile cabalFP $ tfCabal toFetch
 
                 atomically $ modifyTVar outputVar $ Map.insert ident destDir
+            $logWarn $ mconcat $ map (\(path, entryType) -> "Unexpected entry type " <> entryType <> " for entry " <> T.pack path) unexpectedEntries
+
+-- | Internal function used to unpack tarball.
+--
+-- Takes a path to a .tar.gz file, the name of the directory it should contain,
+-- and a destination folder to extract the tarball into. Returns unexpected
+-- entries, as pairs of paths and descriptions.
+untar :: FilePath -> FilePath -> FilePath -> IO [(FilePath, T.Text)]
```

Comment:
I've seen path-io, but this is not trivial since this code has to talk to non-path-ified code.
But you have a point about signatures.

Also, the *existing* code uses FilePath quite a lot, see all the `FP.</>`. I guess you'd be fine with changes to the existing code, both as strictly needed to change `untar`'s signature and as it makes sense? I'll see what I can do later.